tests for parse.py

diivm opened the pull request:
- Using pytest

diivm committed:
init

aPonza commented:
It's a WIP, so I'm not yet sure where everything falls, but try to think smaller, and in terms of expected vs. actual. Moreover a generic test usually runs in "given...when...then" afaik, meaning **given** a struct with `private int foo;` **when** I call parse **then** the json contains a member named `foo` with type `int` and access level `private`  From the input files, it seems like you're trying to test already complex things, while instead I'd expect the tests to be super short, super clear, and super simple.

kunaltyagi commented on bindings/python/tests/input/impl/impl0.hpp:
```diff
@@ -0,0 +1,13 @@
+#ifdef A
```
This is very weird

diivm replied:
removed input files

divmadan commented:
![](https://github.trello.services/images/mini-trello-icon.png) [setup python tests using pytest](https://trello.com/c/dZ2VexVi/29-setup-python-tests-using-pytest)

aPonza commented:
I'd also like to point out you don't necessarily need a real file on the filesystem, if you want you can [mock the lines of a file](https://stackoverflow.com/questions/43567678/python-mock-file-input-for-testing-function) for unit tests. Do what's easier to you.

kunaltyagi commented:
> setup python tests using pytest

Am I missing something here? I don't see any tests (or a link to the trello board if that's what you intended)

> you don't necessarily need a real file on the filesystem

It's better to not need a real filesystem in tests. Using strings (or arrays of strings) also allows the code to be used by more code, removes failure points (works on linux but not on Windows, etc.) and more

divmadan commented:
> Am I missing something here? 

Auto-generated while adding a link in trello.

> It's better to not need a real filesystem in tests.

Test-initiation will start by parsing a header/cpp file. Can we emulate that, without file i/o operations?

aPonza commented:
Sounds like integration testing, I'd like to see unit testing. I'd expect you to `from parse import <function name>`, and in the worst case scenario that function would be what you currently have named `main`, provided you can supply what you'd have as `sys.argv` via mocks (but that's also part of the reason why this was the worst case scenario). In a better scenario, you can provide specific parameters (`compile_commands` and a mocked `source` for `index.parse()` probably) for specific function calls to test specific functions *at their interface*.

You'll quickly see why testing before writing code would have helped you have easily testable code: you could've chosen the way your own code gets to be consumed by a third party, in this case your own testing code.

diivm committed:
minimum

diivm committed:
add test_node_in_this_file

aPonza commented on bindings/python/tests/test_parse.py:
```diff
@@ -0,0 +1,30 @@
+def test_node_in_this_file_false_none():
+    def recursive_call(cursor):
+        for child in cursor.get_children():
+            if child.location.file:
```
I've seen a limited amount of tests in my developer life, but I've never seen branching assertions. You control the *given* portion of the test, so I'd expect you to know exactly what requirements are necessary for your function to return a set value.

diivm replied:
The `if child.location.file` branch removed, as `None` case was removed from `parse.py`: https://github.com/divmadan/pcl/pull/6/commits/2e85ac9ba667459664abea4fe3681e7f1abd2ce0

aPonza commented on bindings/python/tests/test_parse.py:
```diff
@@ -0,0 +1,30 @@
+            if child.location.file:
+                assert parse.node_in_this_file(child, tu.spelling) == False
+            else:
+                assert parse.node_in_this_file(child, tu.spelling) == None
```
Doesn't this function only return boolean values?

diivm replied:
It was returning `None` where `child.location.file` was `None`.
Changed the function in `parse.py`, will change the test too: https://github.com/divmadan/pcl/pull/10/commits/4c5601f17917728e6feacf3c92da4c1c677122cf

Edit:
Change in test: https://github.com/divmadan/pcl/pull/6/commits/2e85ac9ba667459664abea4fe3681e7f1abd2ce0

aPonza reviewed (changes requested):
In general, I don't think the `recursive_call` makes much sense. With this test, you're re-implementing the implementation, and you're testing a function which is only used in the implementation of other functions, so not really at the interface (I'd expect tests for the functions in `main` to make more sense).

divmadan commented:
Needs PR merge: #10

diivm committed:
[test_parse.py] update to remove None

divmadan commented:
> In general, I don't think the `recursive_call` makes much sense. With this test, you're re-implementing the implementation, and you're testing a function which is only used in the implementation of other functions, so not really at the interface (I'd expect tests for the functions in `main` to make more sense).

@aPonza 
Should I not test this function? Without recursive, I'm only testing the node of AST (which will be the filename), not it's children (components).
What do you suggest should be done instead?

aPonza commented:
I don't test functions per se, I test behaviors I want my code to display consistently. Hypothetically, I could have each line of code be its own function, which would mean oh so many tests, if we were testing function-by-function but still only the "important" tests if we were testing the expected behavior of some lines of code.

In this specific case I would've expected a much different experience had the code been written starting with tests. I suppose tiny "integration" tests (Kunal can correct the nomenclature if it's wrong) would do much better. It's like I wrote in https://github.com/divmadan/pcl/pull/6#issuecomment-651151856 but let me expand on it so it's clearer.

**test_member**
Given the unsaved file
```
struct AStruct {
  private int aPrivateIntMember;
};
```
when I parse it, then I expect the output json to `contain` a json object for `aPrivateMember` being an `int` which is `private` and a json object for `AStruct` (optional).

**test_namespace**
Given the unsaved file
```
namespace a_namespace {
struct AStruct {};
} // namespace a_namespace
```
when I parse it, then I expect the output json to `contain` a json object for `a_namespace` which represents a namespace, and a json object for `AStruct` (optional).

**test_include**
Given the source files (notice here I'm back to reasoning with mocks, which I think are more powerful in this case)
*anIncludedFile.hpp*
```
struct AStruct {};
```
and *aFileWhichIncludes.h*
```
#include <anIncludeFile.h>
```
when I parse them, then I expect the output json to `contain` a json object for `AStruct` (mandatory).

I would have expected you to start with these simplistic tests to build and correct the behavior of parse.py so that in the end it parses the whole set of instructions that can happen in the whole library (and therefore a combination of these items). Notice when I say `contain` I have it as code because I expect to be format independent, I just care that it's mentioned. I'm skipping on all the descriptive things you currently have in the json for each node but I hope the bigger picture is clear and that it makes sense.

divmadan commented:
@aPonza 
Thank you so much for the great explanation. Much clearer now, and I get how testing the interface will be more robust in the long term (rather than testing functions).

Just one thing,
> when I parse them, then I expect the output json to contain a json object for AStruct (mandatory).

The include file's nodes will be skipped, according to the current code, i.e. the json will be empty for the file (just translation unit).
The reason for this was that there is no replacement, it parses as every include as is, so it's better to just parse the include file separately and generate it's json, to avoid duplication.

For ex:
Consider a `.h` file contains struct decl: `struct ABC{};`, and it includes a `.hpp` which has it's definition: `struct ABC{int a;}`.

The json will have:
```
<struct ABC> (from .hpp)
    <int a>
<struct ABC> (from .h)
```

kunaltyagi commented:
A few questions:
* How is the code going to know which header to include for what symbol?
* How are you going to handle conflicting json (same symbol in 2 files)? Eg: `namespace detail`

diivm committed:
[test_parse.py] update

diivm committed:
remove __init__.py

diivm committed:
[compile_commands.json] add compile_commands.json for tests

divmadan commented:
I rebased https://github.com/divmadan/pcl/tree/fix/correct_parse into it, so there are some extra commits for now.

diivm committed:
[test_parse.py] type assertion change

diivm committed:
[test_parse.py] add cxx_base_specifier

diivm committed:
[test_parse.py] some renaming

kunaltyagi commented on bindings/python/scripts/context.py:
```diff
@@ -0,0 +1,6 @@
+import os
```
If you're branches are dependent on each other, change the base branch of the PR so the same diff doesn't appear in 2 PRs

kunaltyagi reviewed (commented):
What's new, what's already discussed is hard to tell

diivm committed:
[test_parse.py] remove commented unneeded

diivm committed:
[test_parse.py] use a function for temp file writing and parsing

diivm committed:
[test_parse.py] add tests and refactor

diivm committed:
[test_parse.py] add more tests

diivm committed:
[test_parse.py] add more tests and refactor

diivm committed:
add newline

diivm committed:
[test_parse.py] push for testing

kunaltyagi commented on bindings/python/tests/test_parse.py:
```diff
@@ -0,0 +1,237 @@
+    parsed_info = get_parsed_info(tmp_path=tmp_path, file_contents=file_contents)
+
+    assert type(parsed_info) is dict
+    assert type(parsed_info["members"]) is list
```
It should also be empty

diivm replied:
Done. Good catch!

kunaltyagi commented on bindings/python/tests/test_parse.py:
```diff
@@ -0,0 +1,237 @@
+
+    namespace = parsed_info["members"][0]
+
+    assert namespace["kind"] == "NAMESPACE"
```
Also test the names, for the complete parse check

diivm replied:
Done :+1:

kunaltyagi commented on bindings/python/tests/test_parse.py:
```diff
@@ -0,0 +1,237 @@
+    file_contents = "#include<ostream> \n std::ostream Aostream;"
+    parsed_info = get_parsed_info(tmp_path=tmp_path, file_contents=file_contents)
+
+    namespace_ref = parsed_info["members"][0]["members"][0]
```
Why is this so? Test seems a bit confusing to me. `parsed_info["members"][0]` is the `#include` or the `std::ostream ...;`?

diivm replied:
I forgot to expand this like in other tests. Done :+1:

kunaltyagi commented on bindings/python/tests/test_parse.py:
```diff
@@ -0,0 +1,237 @@
+
+
+def test_struct_decl(tmp_path):
+    file_contents = "struct AStruct{};"
```
Better names like `EmptyStruct` or `DeclOnly` (not this one), etc. would be nice

diivm replied:
I followed this pattern: `test_<cursor_kind>`, because we are testing that entity.

> EmptyStruct or DeclOnly

Separate tests for such cases should be added IMO, keeping the cursor kind checks separate.

kunaltyagi replied:
I was referring to the names in file content

diivm replied:
Updated as per @aponza's recommendations.

kunaltyagi commented on bindings/python/tests/test_parse.py:
```diff
@@ -0,0 +1,237 @@
+    file_contents = "struct AStruct{}; struct BStruct: public AStruct{};"
+    parsed_info = get_parsed_info(tmp_path=tmp_path, file_contents=file_contents)
+
+    child_struct_decl = parsed_info["members"][1]
```
A rephrasing would be better
```py
for member in parsed_info['members']:
  if member['name'] == 'Derived':
    child_struct_decl = member
    break
```

Try to show the assumption without needing to make leaps of faith

diivm replied:
That was my initial approach, but @aponza advised to just use index since we know exactly how it has to be. I am in favour of the `for loop` approach, even though libclang is pretty stable.

Some input from @aponza?

aPonza replied:
I try to follow the best practice of not testing the implementation, be it yours or theirs. In this case a `contains`+`get` would do wonders, you can implement it with Kunal's for loop or with something like [`any`](https://thispointer.com/python-how-to-check-if-an-item-exists-in-list-search-by-value-or-condition/): `child_struct_decl = any(member['name'] == 'Derived' for member in parsed_info['members']`) should be the line, right? EDIT: nope, any returns a boolean, probably use Kunal's.

Also note I'd change `AStruct` - `BStruct`to `ABaseStruct`/`AParentStruct` - `ADerivedStruct`/`AChildStruct` and in general, the prefixing a/an is to be more descriptive of what thing I'm using. I don't know if it's a widespread practice, it's what I do myself, so variables could be `a_very_long_file_hash`, `expected_json`, ...

kunaltyagi replied:
I'd definitely prefer `ADerivedStruct` over `Derived` (my example). My reasoning for using a loop:
* I don't know if the implementing library guarantees ordered AST

If `clang` guarantees that, then no need and my suggestion is verbose because I'm new to the library

diivm replied:
> If clang guarantees that

How do we check that?

Here's the function in clang:
```py
    def get_children(self):
        """Return an iterator for accessing the children of this cursor."""

        def visitor(child, parent, children):
            assert child != conf.lib.clang_getNullCursor()

            # Create reference to TU so it isn't GC'd before Cursor.
            child._tu = self._tu
            children.append(child)
            return 1 # continue
        children = []
        conf.lib.clang_visitChildren(self, callbacks['cursor_visit'](visitor),
            children)
        return iter(children)
```

kunaltyagi replied:
Dunno. Maybe note the assumption in case the test fails in future?

kunaltyagi commented on bindings/python/tests/test_parse.py:
```diff
@@ -0,0 +1,237 @@
+    file_contents = "struct AStruct{void Afunction();}"
+    parsed_info = get_parsed_info(tmp_path=tmp_path, file_contents=file_contents)
+
+    struct_decl = parsed_info["members"][0]
```
Assert that the struct_decl is indeed the struct decl. While reading, it should go like:
1. get the first member
2. it's the struct you just read above
3. get it's first member
4. it's a method
5. it has a name: same as before
6. it even has the same return type and parameters
7. QED

diivm replied:
We can do that, but the member and function, etc. are getting tested in another function. 
Following @aponza's **Tests are like linear combinations** analogy.

@aponza, some input?

aPonza replied:
My point was to avoid redundant tests, but readability is also key. There's no objective way to know if something is readable, but Kunal's proposal is at least as readable as mine.

kunaltyagi replied:
With the additional tests on `Afunction`, I'd say skip step 2. Issue resolved from my side :)

diivm committed:
[test_parse.py] add empty condition check

diivm committed:
[test_parse.py] updates and fixes

- change AStruct to Astruct (and B) for uniformity
- add name testing
- expand namespace_ref test's parsed_info

aPonza commented on bindings/python/tests/test_parse.py:
```diff
@@ -0,0 +1,269 @@
+
+
+def test_cxx_method(tmp_path):
+    file_contents = "struct Astruct{void Afunction();}"
```
capitalization for C++ mandates `AStruct` and `aFunction`.

aPonza commented on bindings/python/tests/test_parse.py:
```diff
@@ -0,0 +1,269 @@
+
+    assert cxx_base_specifier["kind"] == "CXX_BASE_SPECIFIER"
+    assert cxx_base_specifier["access_specifier"] == "PUBLIC"
+    assert cxx_base_specifier["name"] == "struct Astruct"
```
capitalization for C++ mandates `AStruct`

aPonza commented on bindings/python/tests/test_parse.py:
```diff
@@ -0,0 +1,269 @@
+    ]
+    input.write_bytes(str(x).encode())
+    tmp_path = str(tmp_path)
+    return tmp_path
```
You can simplify this with `with open`. You can also skip naming the variable you return in most cases where you do (so `return str(tmp_path)` unless it's unreadable as is.

aPonza commented on bindings/python/tests/test_parse.py:
```diff
@@ -0,0 +1,269 @@
+
+def get_parsed_info(tmp_path, file_contents):
+    source_path = tmp_path / "file.hpp"
+    source_path.write_bytes(str(file_contents).encode())
```
Simplify with `with open`

aPonza commented on bindings/python/tests/test_parse.py:
```diff
@@ -0,0 +1,269 @@
+    return parsed_info
+
+
+def test_types(tmp_path):
```
Maybe `test_return_type_composition` or something along that line makes more sense?

aPonza commented on bindings/python/tests/test_parse.py:
```diff
@@ -0,0 +1,269 @@
+
+
+def test_namespace(tmp_path):
+    file_contents = "namespace Anamespace {}"
```
capitalization for C++ mandates `a_namespace`

aPonza commented on bindings/python/tests/test_parse.py:
```diff
@@ -0,0 +1,269 @@
+
+
+def test_namespace_ref(tmp_path):
+    file_contents = "#include<ostream> \n std::ostream Aostream;"
```
Capitalization for C++ mandates `anOstream`. Also reminder to implement what Sergio suggested on discord

aPonza commented on bindings/python/tests/test_parse.py:
```diff
@@ -0,0 +1,269 @@
+
+
+def test_struct_decl(tmp_path):
+    file_contents = "struct Astruct{};"
```
Capitalization for C++ mandates `AStruct`.

aPonza commented on bindings/python/tests/test_parse.py:
```diff
@@ -0,0 +1,269 @@
+
+
+def test_var_decl(tmp_path):
+    file_contents = "int Aint=1;"
```
capitalization for C++ mandates `anInt`

aPonza commented on bindings/python/tests/test_parse.py:
```diff
@@ -0,0 +1,269 @@
+    var_decl = parsed_info["members"][0]
+
+    assert var_decl["kind"] == "VAR_DECL"
+    assert var_decl["element_type"] == "Int"
```
does this actually pass? shouldn't this be "**i**nt"?

diivm replied:
Yes. libclang parses element type as Int, Float, etc.

aPonza commented on bindings/python/tests/test_parse.py:
```diff
@@ -0,0 +1,269 @@
+
+
+def test_type_ref(tmp_path):
+    file_contents = "struct Astruct{int Aint;} \n void Afunction(Astruct& Aint);"
```
capitalization for C++ mandates `AStruct`, `anInt`, `aFunction`, ... Also implement Sergio's suggestion from discord

aPonza commented on bindings/python/tests/test_parse.py:
```diff
@@ -0,0 +1,269 @@
+
+
+def test_constructor(tmp_path):
+    file_contents = "struct Astruct{Astruct(){}};"
```
Capitalization for C++ mandates `AStruct`.

aPonza commented on bindings/python/tests/test_parse.py:
```diff
@@ -0,0 +1,269 @@
+
+
+def test_parm_decl(tmp_path):
+    file_contents = "struct Astruct{ Astruct(int Aint) {} };"
```
Capitalization for C++ mandates `AStruct`, `anInt`.

aPonza commented on bindings/python/tests/test_parse.py:
```diff
@@ -0,0 +1,269 @@
+    parm_decl = constructor["members"][0]
+
+    assert parm_decl["kind"] == "PARM_DECL"
+    assert parm_decl["element_type"] == "Int"
```
Should this be Int with capital I or int?

diivm replied:
Int.

aPonza commented on bindings/python/tests/test_parse.py:
```diff
@@ -0,0 +1,269 @@
+
+
+def test_call_expr(tmp_path):
+    file_contents = "struct Astruct{ Astruct (): Astruct() {} };"
```
Capitalization for C++ mandates `AStruct`. Also if you use the `"""` suggestion from Sergio you could make this more readable (usual style, I mean) with some line breaks

aPonza commented on bindings/python/tests/test_parse.py:
```diff
@@ -0,0 +1,269 @@
+
+def test_unexposed_expr(tmp_path):
+    file_contents = "struct Astruct{ int Aint; \n\
+        Astruct (int Bint): Aint(Bint) {} };"
```
Capitalization for C++ mandates `AStruct`, `aClassMember` and `aFunctionArgument` (?) Also use vertical space as per what a C++ file would look like.

aPonza commented on bindings/python/tests/test_parse.py:
```diff
@@ -0,0 +1,269 @@
+
+def test_decl_ref_expr(tmp_path):
+    file_contents = "struct Astruct{ int Aint; \n\
+        Astruct (int Bint): Aint(Bint) {} };"
```
same

aPonza commented on bindings/python/tests/test_parse.py:
```diff
@@ -0,0 +1,269 @@
+
+
+def test_field_decl(tmp_path):
+    file_contents = "struct Astruct{ int Aint; };"
```
same

aPonza commented on bindings/python/tests/test_parse.py:
```diff
@@ -0,0 +1,269 @@
+
+def test_member_ref(tmp_path):
+    file_contents = "struct Astruct{ int Aint; \n\
+        Astruct (int Bint): Aint(Bint) {} };"
```
same

aPonza commented on bindings/python/tests/test_parse.py:
```diff
@@ -0,0 +1,269 @@
+
+
+def test_class_template(tmp_path):
+    file_contents = "template<typename N> struct Astruct {};"
```
same

aPonza commented on bindings/python/tests/test_parse.py:
```diff
@@ -0,0 +1,269 @@
+
+
+def test_template_non_type_parameter(tmp_path):
+    file_contents = "template<int N> struct Astruct {};"
```
same

aPonza commented on bindings/python/tests/test_parse.py:
```diff
@@ -0,0 +1,269 @@
+
+
+def test_function_template(tmp_path):
+    file_contents = "template<int N> void Afunction() {}"
```
same

aPonza commented on bindings/python/tests/test_parse.py:
```diff
@@ -0,0 +1,269 @@
+
+
+def test_template_type_parameter(tmp_path):
+    file_contents = "template<typename N> struct Astruct {};"
```
same

aPonza reviewed (commented):
In general, the names you give to variables and classes in mock files should help the reader understand the single test you're performing, so if an integer in that context might represent an R-value reference or what have you, name it in such a way it's understandable, as you would for a variable in real code. I'd like to know from Kunal if the general structure of the tests in this file makes sense to him as well: to me the objective is to show all the possible outcomes from parsing every possible file, in a linearly combinable fashion, meaning all possible `kind`s are explicated via dedicated tests. My ideal setup would again have been with a "contains"-based check and not an equality-based check. An itsy-bitsy-integration test down the line could show all these `kind`s work together, but that might be very redundant if there's no code required to make it pass.

diivm committed:
[test_parse.py] file i/o, refactoring and multi line str

kunaltyagi commented on bindings/python/scripts/parse.py:
```diff
@@ -121,23 +121,23 @@ def generate_parsed_info(node):
-            parsed_info["raw_comment"] = cursor.raw_comment
-        parsed_info["members"] = []
+    # if cursor.spelling != "":
+    parsed_info["depth"] = depth
```
Change base of PR. Why are these changes showing up in this PR? :confused:

diivm replied:
Oops, needed a rebase.

kunaltyagi commented on bindings/python/tests/test_parse.py:
```diff
@@ -0,0 +1,332 @@
+    assert var_decl["name"] == "anInt"
+
+
+def test_type_ref(tmp_path):
```
This test could be simpler, or else the name is wrong

diivm replied:
Something like this?
```cpp
    typedef size_type;
    size_type size;
```

kunaltyagi replied:
o.O Is that valid C++?

diivm replied:
Sorry, this:
```cpp
    typedef unsigned size_type;
    size_type size;
```

kunaltyagi replied:
Is that a type ref? Or is a type-ref a type reference (as in `int&` and not just `int`)?

diivm replied:
From cindex:
```py
# A reference to a type declaration.
#
# A type reference occurs anywhere where a type is named but not
# declared. For example, given:
#   typedef unsigned size_type;
#   size_type size;
#
# The typedef is a declaration of size_type (CXCursor_TypedefDecl),
# while the type of the variable "size" is referenced. The cursor
# referenced by the type of size is the typedef for size_type.
CursorKind.TYPE_REF = CursorKind(43)
```

kunaltyagi commented on bindings/python/tests/test_parse.py:
```diff
@@ -0,0 +1,332 @@
+    assert parm_decl["name"] == "aFunctionArgument"
+
+
+def test_call_expr(tmp_path):
```
Maybe have a different example here. This one is good, but too many AStruct to confuse a reader.
```py
struct AStruct {};
struct BStruct { BStruct(): AStruct() {} };
```

diivm replied:
Something without struct, perhaps?
```cpp
    int aFunction() {
        return 1;
    }
    int anInt = aFunction();
```

kunaltyagi commented on bindings/python/tests/test_parse.py:
```diff
@@ -0,0 +1,332 @@
+    file_contents = """
+    struct AStruct {
+        int aClassMember;
+        AStruct(int aFunctionArgument) : aClassMember(aFunctionArgument) {}
```
This example raises questions regarding:
```py
AStruct(int firstArg, int secondArg): firstMember(secondArg), secondMember(firstArg) {}
```

diivm replied:
It works.
Add this one too?

kunaltyagi replied:
The question is would `secondMember` be the 3rd element or the 2nd (aka index 2 or 1)?

diivm replied:
```diff
- ...
                        {
                            "depth": 3,
                            "line": 4,
                            "column": 86,
                            "kind": "UNEXPOSED_EXPR",
                            "name": "secondFunctionArgument",
                            "element_type": "Int",
                            "members": [
                                {
                                    "depth": 4,
                                    "line": 4,
                                    "column": 86,
                                    "kind": "DECL_REF_EXPR",
                                    "name": "secondFunctionArgument",
                                    "element_type": "Int",
                                    "members": [],
                                }
                            ],
                        },
- ...
                        {
                            "depth": 3,
                            "line": 4,
                            "column": 124,
                            "kind": "UNEXPOSED_EXPR",
                            "name": "firstFunctionArgument",
                            "element_type": "Int",
                            "members": [
                                {
                                    "depth": 4,
                                    "line": 4,
                                    "column": 124,
                                    "kind": "DECL_REF_EXPR",
                                    "name": "firstFunctionArgument",
                                    "element_type": "Int",
                                    "members": [],
                                }
- ...
```

This should be tested at decl_ref_expr, instead of unexposed_expr
Something like this:

```py
def test_decl_ref_expr(tmp_path):
    file_contents = """
    struct AStruct { 
        int firstMember, secondMember;
        AStruct(int firstFunctionArgument, int secondFunctionArgument) : firstMember(secondFunctionArgument), secondMember(firstFunctionArgument) {}
    };
    """
    parsed_info = get_parsed_info(tmp_path=tmp_path, file_contents=file_contents)

    struct_decl = parsed_info["members"][0]
    constructor = struct_decl["members"][2]
    unexposed_expr_1 = constructor["members"][3]
    unexposed_expr_2 = constructor["members"][5]
    decl_ref_expr_1 = unexposed_expr_1["members"][0]
    decl_ref_expr_2 = unexposed_expr_2["members"][0]

    assert decl_ref_expr_1["kind"] == "DECL_REF_EXPR"
    assert decl_ref_expr_2["kind"] == "DECL_REF_EXPR"
    assert decl_ref_expr_1["name"] == "secondFunctionArgument"
    assert decl_ref_expr_2["name"] == "firstFunctionArgument"
```

For member ref:
```py
def test_member_ref(tmp_path):
    file_contents = """
    struct AStruct { 
        int firstMember, secondMember;
        AStruct(int firstFunctionArgument, int secondFunctionArgument) : firstMember(secondFunctionArgument), secondMember(firstFunctionArgument) {}
    };
    """
    parsed_info = get_parsed_info(tmp_path=tmp_path, file_contents=file_contents)
    struct_decl = parsed_info["members"][0]
    constructor = struct_decl["members"][2]
    member_ref_1 = constructor["members"][2]
    member_ref_2 = constructor["members"][4]

    assert member_ref_1["kind"] == "MEMBER_REF"
    assert member_ref_2["kind"] == "MEMBER_REF"
    assert member_ref_1["element_type"] == "Int"
    assert member_ref_2["element_type"] == "Int"
    assert member_ref_1["name"] == "firstMember"
    assert member_ref_2["name"] == "secondMember"
```

kunaltyagi reviewed (commented):
only viewed a few tests

divmadan commented:
Closing this in favour of #18, #19, #20
